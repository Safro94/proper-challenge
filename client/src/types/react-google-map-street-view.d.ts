declare module 'react-google-map-street-view';
