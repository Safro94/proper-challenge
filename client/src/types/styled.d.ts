import 'styled-components';

interface IPalette {
	main: string;
	text: string;
}

interface IBreakpoints {
	mobile: string;
	tablet: string;
	laptop: string;
	desktop: string;
}

declare module 'styled-components' {
	export interface DefaultTheme {
		breakpoints: IBreakpoints;
		palette: {
			common: {
				black: string;
				gray: string;
				white: string;
				danger: string;
			};
			primary: IPalette;
		};
	}
}
