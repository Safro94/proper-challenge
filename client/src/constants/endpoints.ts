export const PROPERTIES_ENDPOINT: string = '/api/properties';
