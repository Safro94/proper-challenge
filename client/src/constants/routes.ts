export const HOME = '/';
export const PROPERTY_DETAIL = '/property/:propertyId';
export const NEW_PROPERTY = '/property/add';
