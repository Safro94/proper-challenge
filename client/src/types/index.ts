export interface IProperty {
	id: number;
	address: string;
	size: string;
	rooms: number;
	utilities: string;
	tenantName: string;
}

export enum ButtonTypes {
	Primary = 'primary',
	Danger = 'danger',
}

export enum RequestStatus {
	Idle = 'idle',
	Pending = 'pending',
	Resolved = 'resolved',
	Rejected = 'rejected',
}
