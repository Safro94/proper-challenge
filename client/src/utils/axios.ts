import axios, { AxiosError, AxiosRequestConfig } from 'axios';
import https from 'https';

const opts: AxiosRequestConfig = {
	httpsAgent: new https.Agent({ rejectUnauthorized: false }),
};

const interceptorError = (error: AxiosError) => Promise.reject(error.response);

const instance = axios.create(opts);

instance.interceptors.response.use(response => response, interceptorError);
instance.defaults.headers.common['x-api-key'] = process.env.REACT_APP_API_KEY;

export default instance;
