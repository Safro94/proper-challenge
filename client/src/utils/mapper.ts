import { IAddress, IDawaResponse } from '../types';

const mapDawaResponse = (data: IDawaResponse[]): IAddress[] => {
	return data.map(item => ({
		text: item.tekst,
		type: item.type,
	}));
};

export { mapDawaResponse };
