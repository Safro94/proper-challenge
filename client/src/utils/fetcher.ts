import { RequestMethods, RequestStatus } from '../types';

import axios from './axios';

interface IFetcher {
	method?: RequestMethods;
	url: string;
	data?: null | any;
	body?: null | any;
	headers?: null | any;
}

interface IResponse {
	data: any;
	status: RequestStatus;
}

const fetcher = async ({
	method = RequestMethods.GET,
	url,
	body = null,
	headers = null,
}: IFetcher): Promise<IResponse> => {
	let data;
	let status = RequestStatus.Pending;

	try {
		const result = await axios[method](
			url,
			JSON.parse(headers),
			JSON.parse(body)
		);
		data = result.data;
		status = RequestStatus.Resolved;
	} catch (error) {
		status = RequestStatus.Rejected;
		throw error;
	}

	return { data, status };
};

export default fetcher;
