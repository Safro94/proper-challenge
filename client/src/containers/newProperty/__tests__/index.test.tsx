import { render, screen, waitFor } from '../../../utils/test-utils';

import NewPropertyContainer from '..';

import fetcher from '../../../utils/fetcher';
import { PROPERTIES_ENDPOINT } from '../../../constants/endpoints';
import { IAddress, RequestMethods, RequestStatus } from '../../../types';
import userEvent from '@testing-library/user-event';
import { HOME } from '../../../constants/routes';

jest.mock('../../../utils/fetcher');

jest.mock('../../../components/loading', () => () => <div>Loading</div>);

const mockHistoryPush = jest.fn();
jest.mock('react-router-dom', () => ({
	...jest.requireActual('react-router-dom'),
	useHistory: () => ({
		push: mockHistoryPush,
	}),
}));

const fetchData = [
	{ type: 'address', tekst: 'aa 35' },
	{ type: 'address', tekst: 'aa 40' },
	{ type: 'address', tekst: 'aa 40a' },
	{ type: 'address', tekst: 'ab 30' },
	{ type: 'address', tekst: 'ac 50' },
];

describe('NewPropertyContainer', () => {
	it('should render correctly', async () => {
		const { container } = render(<NewPropertyContainer />);

		await waitFor(() => {
			expect(container.firstChild).toMatchSnapshot();
		});
	});

	it('should show a disabled button', () => {
		render(<NewPropertyContainer />);

		expect(screen.getByRole('button', { name: /add/i })).toBeDisabled();
	});

	it('should show the addresses in the dropdown, filter while the user types and call the fetcher', async () => {
		(fetcher as jest.Mock).mockImplementationOnce(() =>
			Promise.resolve({ data: fetchData, status: RequestStatus.Resolved })
		);

		(fetcher as jest.Mock).mockImplementationOnce(() =>
			Promise.resolve({
				data: fetchData.filter(item => item.tekst.includes('aa')),
				status: RequestStatus.Resolved,
			})
		);

		(fetcher as jest.Mock).mockImplementationOnce(() =>
			Promise.resolve({
				data: fetchData.filter(item => item.tekst.includes('aa 40')),
				status: RequestStatus.Resolved,
			})
		);

		(fetcher as jest.Mock).mockImplementationOnce(() => Promise.resolve());

		render(<NewPropertyContainer />);

		expect(fetcher).toHaveBeenCalledTimes(0);

		const addButton = screen.getByRole('button', { name: /add/i });
		expect(addButton).toBeDisabled();

		const editButton = screen.getByRole('button', { name: /edit/i });
		expect(editButton).toBeDisabled();

		const input = screen.getByPlaceholderText(/enter address/i);

		await waitFor(() => {
			userEvent.type(input, 'a');
			expect(fetcher).toHaveBeenCalledTimes(1);
			expect((fetcher as jest.Mock).mock.calls[0][0]).toStrictEqual({
				url: `${process.env.REACT_APP_DAWA_API_ENDPOINT}a&caretpos=18&fuzzy=`,
			});
		});

		await waitFor(() => {
			userEvent.type(input, 'a');
			expect(fetcher).toHaveBeenCalledTimes(2);
			expect((fetcher as jest.Mock).mock.calls[1][0]).toStrictEqual({
				url: `${process.env.REACT_APP_DAWA_API_ENDPOINT}aa&caretpos=18&fuzzy=`,
			});
		});

		expect(screen.getAllByRole('option')).toHaveLength(3);
		const selectedOption = screen.getByRole('option', { name: 'aa 40' });
		userEvent.click(selectedOption);
		expect(editButton).not.toBeDisabled();

		userEvent.type(screen.getByPlaceholderText(/enter tenant/i), 'M');
		userEvent.type(screen.getByPlaceholderText(/enter room/i), '2');
		userEvent.type(screen.getByPlaceholderText(/enter size/i), '2');
		userEvent.type(screen.getByPlaceholderText(/enter utilities/i), 'c');
		expect(addButton).not.toBeDisabled();

		userEvent.click(editButton);
		expect(addButton).toBeDisabled();

		userEvent.type(input, 'a');

		await waitFor(() => {
			expect(fetcher).toHaveBeenCalledTimes(3);
			expect((fetcher as jest.Mock).mock.calls[2][0]).toStrictEqual({
				url: `${process.env.REACT_APP_DAWA_API_ENDPOINT}aa 40a&caretpos=18&fuzzy=`,
			});
		});

		const newSelectedOption = screen.getByRole('option', { name: /aa 40a/i });
		userEvent.click(newSelectedOption);

		expect(addButton).not.toBeDisabled();

		userEvent.click(addButton);
		expect(fetcher).toHaveBeenCalledTimes(4);
		expect((fetcher as jest.Mock).mock.calls[3][0]).toStrictEqual({
			url: `${process.env.REACT_APP_SERVER_URL}${PROPERTIES_ENDPOINT}`,
			method: RequestMethods.POST,
			body: {
				address: 'aa 40a',
				rooms: 12,
				size: '2',
				tenantName: 'M',
				utilities: 'c',
			},
		});

		await waitFor(() => {
			expect(mockHistoryPush).toHaveBeenCalledTimes(1);
			expect(mockHistoryPush).toHaveBeenCalledWith(HOME);
		});
	});
});
