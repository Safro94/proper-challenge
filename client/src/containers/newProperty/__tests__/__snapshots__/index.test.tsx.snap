// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewPropertyContainer should render correctly 1`] = `
<div
  class="sc-gKAaRy clasGk"
>
  <h2
    class="sc-crzoAE jfCOzI"
  >
    newPropertyContainer:newPropertyTitle
  </h2>
  <form
    class="sc-jrsJWt jKmbSQ"
    method="POST"
  >
    <div
      class="sc-dIsUp kQIqJD"
    >
      <div
        class="sc-ksluID bdaegv"
      >
        <div
          class="sc-bdnxRM iWitlZ"
        >
          <div
            class="sc-dlnjwi cSrDLl"
          >
            <label
              class="sc-eCApnc fahIwV"
              for="downshift-0-input"
              id="downshift-0-label"
            >
              address
            </label>
            <div
              aria-expanded="false"
              aria-haspopup="listbox"
              aria-owns="downshift-0-menu"
              class="sc-gtsrHT gIlwKp"
              role="combobox"
            >
              <input
                aria-autocomplete="list"
                aria-controls="downshift-0-menu"
                aria-labelledby="downshift-0-label"
                autocomplete="off"
                class="sc-hKFxyN daVEHi"
                id="downshift-0-input"
                placeholder="newPropertyContainer:enter address"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="sc-bqGGPW grIKgg"
      >
        <button
          class="sc-jSFjdj bfWMye"
          disabled=""
          type="button"
        >
          edit
        </button>
      </div>
    </div>
    <div
      class="sc-ksluID kJcpKk"
    >
      <label
        class="sc-kEqXSa gsioAq"
      >
        tenant
      </label>
      <input
        class="sc-iqAclL kVfQdq"
        placeholder="newPropertyContainer:enter tenant"
        type="text"
        value=""
      />
    </div>
    <div
      class="sc-ksluID kJcpKk"
    >
      <label
        class="sc-kEqXSa gsioAq"
      >
        rooms
      </label>
      <input
        class="sc-iqAclL kVfQdq"
        placeholder="newPropertyContainer:enter rooms"
        type="number"
        value="1"
      />
    </div>
    <div
      class="sc-ksluID kJcpKk"
    >
      <label
        class="sc-kEqXSa gsioAq"
      >
        size
      </label>
      <input
        class="sc-iqAclL kVfQdq"
        placeholder="newPropertyContainer:enter size"
        type="text"
        value=""
      />
    </div>
    <div
      class="sc-ksluID kJcpKk"
    >
      <label
        class="sc-kEqXSa gsioAq"
      >
        utilities
      </label>
      <input
        class="sc-iqAclL kVfQdq"
        placeholder="newPropertyContainer:enter utilities"
        type="text"
        value=""
      />
    </div>
    <button
      class="sc-jSFjdj bfWMye"
      disabled=""
      type="submit"
    >
      add
    </button>
  </form>
</div>
`;
